examples/migrating_trench/trench_example.py

```diff
@@ -0,0 +1,168 @@
+"""
+Migrating Trench Test case
+=======================
+
+Solves the test case of a migrating trench.
+
+We use this test case to validate the implementation of the mathematical and
+numerical methods used in Thetis to model sediment transport and morphological changes.
+In the figure produced, we compare our results with experimental data from a lab study
+
+For more details, see
+[1] Clare et al. 2020. “Hydro-morphodynamics 2D Modelling Using a Discontinuous
+    Galerkin Discretisation.” EarthArXiv. January 9. doi:10.31223/osf.io/tpqvy.
+
+"""
+
+from thetis import *
+
+import numpy as np
+import pandas as pd
+import pylab as plt
+import time
+
+conservative = False
+
+# Note it is necessary to run trench_hydro first to get the hydrodynamics simulation
+
+# define mesh
+lx = 16
+ly = 1.1
+nx = lx*5
+ny = 5
+mesh2d = RectangleMesh(nx, ny, lx, ly)
+
+x, y = SpatialCoordinate(mesh2d)
+
+# define function spaces
+V = FunctionSpace(mesh2d, "CG", 1)
+DG_2d = FunctionSpace(mesh2d, "DG", 1)
+vector_dg = VectorFunctionSpace(mesh2d, "DG", 1)
+
+# define underlying bathymetry
+bathymetry_2d = Function(V, name='bathymetry_2d')
+initialdepth = Constant(0.397)
+depth_riv = Constant(initialdepth - 0.397)
+depth_trench = Constant(depth_riv - 0.15)
+depth_diff = depth_trench - depth_riv
+
+trench = conditional(le(x, 5), depth_riv, conditional(le(x, 6.5), (1/1.5)*depth_diff*(x-6.5) + depth_trench,
+                     conditional(le(x, 9.5), depth_trench, conditional(le(x, 11), -(1/1.5)*depth_diff*(x-11) + depth_riv, depth_riv))))
+bathymetry_2d.interpolate(-trench)
+
+# choose directory to output results
+ts = time.time()
+st = datetime.datetime.fromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S')
+outputdir = 'outputs' + st
+
+print_output('Exporting to '+outputdir)
+
+# define bathymetry_file
+bathy_file = File(outputdir + "/bathy.pvd")
+
+morfac = 100
+dt = 0.3
+end_time = 15*3600
+
+diffusivity = 0.15
+viscosity_hydro = Constant(1e-6)
+
+# initialise velocity and elevation
+chk = DumbCheckpoint("hydrodynamics_trench/elevation", mode=FILE_READ)
```

Comment:
you can use `load_state` to read in the initial conditions from a different directory. at the moment it only reads the hydro state, not tracers

Reply:
also, you provide the hydro spin up script as well as the spun-up solution as h5 files. Would the former be sufficient?

Reply:
I agree as the h5 files are probably going to break at some point. I actually removed these previously, but it seems they've been added back in some local merge.
Re. using `load_state`: hm, that seems to be generally broken at the moment `load_state` never loads any tracer (or sediment fields)! In the case of exner I guess it should also read a bathymetry checkpoint. That brings however another issue: the spin up model does not have any tracers, so if we call load_state() here it would error.  This might be something that is quite common with tracers in general: spinning up the hydrodynamics to a checkpoint and restart (release) from a user provided tracer (unlike temperature+salinity in 3d that are usually more tightly coupled with dynamics). So maybe in this case it makes more sense to *not* use `load_state`? Note in these cases, we also don't really care about restarting exactly at the same time-level and contueing the indexing of the output files, rather you just want to start at t=0 and index=0.
Regardless, we should really fix checkpointing of 2d runs with tracer and/or sediment when it's really used as a restart, maybe in a separate PR?

Reply:
Yes, restarting is currently broken for tracers. I agree that we should fix it in another PR. We could plausibly add an option to `load_state` to only load the hydrodynamics. Also, I think if you override `t=0` and `iteration=0` you can just load the init state without setting the index. Could make the API cleaner though.